src/descriptor/key.rs

```diff
@@ -525,42 +541,167 @@ impl error::Error for ConversionError {
+
+    /// Whether or not this key has multiple derivation paths.
+    fn is_multipath(&self) -> bool;
+}
+
+impl DescriptorInnerKey for DescriptorSinglePublicKey {
+    fn master_fingerprint(&self) -> bip32::Fingerprint {
+        if let Some((fingerprint, _)) = self.origin {
+            fingerprint
+        } else {
+            let mut engine = XpubIdentifier::engine();
+            match self.key {
+                SinglePubKey::FullKey(pk) => {
+                    pk.write_into(&mut engine).expect("engines don't error")
+                }
+                SinglePubKey::XOnly(x_only_pk) => engine.input(&x_only_pk.serialize()),
+            };
+            bip32::Fingerprint::from(
+                &XpubIdentifier::from_engine(engine)[..4]
+                    .try_into()
+                    .expect("4 byte slice"),
+            )
+        }
+    }
+
+    fn full_derivation_path(&self) -> Option<bip32::DerivationPath> {
+        Some(if let Some((_, ref path)) = self.origin {
+            path.clone()
+        } else {
+            bip32::DerivationPath::from(vec![])
+        })
+    }
+
+    fn has_wildcard(&self) -> bool {
+        false
+    }
+
+    fn at_derivation_index(self, _index: u32) -> Result<DefiniteDescriptorKey, ConversionError> {
+        Ok(DefiniteDescriptorKey::new(DescriptorPublicKey::Single(self))
+            .expect("The key should not contain any wildcards at this point"))
+    }
+
+    fn is_multipath(&self) -> bool {
+        false
+    }
+}
+
+impl DescriptorInnerKey for DescriptorExtendedPublicKey {
+    fn master_fingerprint(&self) -> bip32::Fingerprint {
+        if let Some((fingerprint, _)) = self.origin {
+            fingerprint
+        } else {
+            self.xkey.fingerprint()
+        }
+    }
+
+    fn full_derivation_path(&self) -> Option<bip32::DerivationPath> {
+        let origin_path = if let Some((_, ref path)) = self.origin {
+            path.clone()
+        } else {
+            bip32::DerivationPath::from(vec![])
+        };
+        Some(origin_path.extend(&self.derivation_path))
+    }
+
+    fn has_wildcard(&self) -> bool {
+        self.wildcard != Wildcard::None
+    }
+
+    fn at_derivation_index(self, index: u32) -> Result<DefiniteDescriptorKey, ConversionError> {
+        let derivation_path = match self.wildcard {
+            Wildcard::None => self.derivation_path,
+            Wildcard::Unhardened => self.derivation_path.into_child(
+                bip32::ChildNumber::from_normal_idx(index)
+                    .ok()
+                    .ok_or(ConversionError::HardenedChild)?,
+            ),
+            Wildcard::Hardened => self.derivation_path.into_child(
+                bip32::ChildNumber::from_hardened_idx(index)
+                    .ok()
+                    .ok_or(ConversionError::HardenedChild)?,
+            ),
+        };
+        let definite = DescriptorPublicKey::XPub(DescriptorXKey {
```

Comment:
This is an awkward "upward" instantiation of the `DescriptorPublicKey` type. An idea:

1. `DefiniteDescriptorKey::new` could take a `DescriptorInnerKey`
2. `DescriptorInnerKey` extracts and implements `derive_public_key`, which is also a `match` block across the inner enum.

Reply:
Maybe `DescriptorInnerKey` should have an `Into<DescriptorPublicKey>` bound?

Reply:
🧠 💡

Reply:
Erm, that'll require ~`DescriptorInnerKey`~`DescriptorKey` to be made object-safe. That's beyond my ken.

Reply:
I think it might be as simple as sticking a `: Sized` bound on the trait? What error do you get?